conan/api/subapi/report.py

```diff
@@ -0,0 +1,171 @@
+import base64
+import os
+
+from conan.api.output import ConanOutput
+from conan.errors import ConanException
+from conan.api.model import RecipeReference
+from conan.internal.conan_app import ConanApp
+from conan.internal.errors import conanfile_exception_formatter
+from conan.internal.graph.graph import CONTEXT_HOST
+from conan.internal.graph.profile_node_definer import initialize_conanfile_profile
+from conan.internal.source import config_source
+
+
+class ReportAPI:
+    def __init__(self, conan_api):
+        self.conan_api = conan_api
+
+    def diff(self, old_reference, new_reference, remotes, old_path=None, new_path=None, cwd=None):
+        """
+        Compare two recipes and return the differences.
+        :param old_reference: The reference of the old recipe.
+        :param new_reference: The reference of the new recipe.
+        :param remotes: List of remotes to search for the recipes.
+        :param old_path: Optional path to the old recipe's conanfile.py.
+        :param new_path: Optional path to the new recipe's conanfile.py.
+        :param cwd: Current working directory, used to resolve paths.
+        :return: A dictionary with the differences between the two recipes.
+        """
+
+        def _source(path_to_conanfile, reference):
+            if path_to_conanfile is None:
+                export_ref, cache_path = _download_ref_from_remote(self.conan_api, reference, remotes)
```

Comment:
And what if the reference is already in the cache, and doesn't need to be downloaded?

Reply:
Download ref now first looks in the cache, this will make it behave the same as the rest of the conan commands that first try to resolve in the cache

(but the feature might still be off, need to double check)

Reply:
Maybe just rename the function to something like ``get_ref_from_cache_or_remote()``? Or something shorter, but that doesn't look like that it will always download.